### PR TITLE
test/extended/prometheus: Add test for no alerts in pending state

### PR DESCRIPTION
**Pull request by lilic**
Because our tests run only for short amount of time they might not
catch the alerts that start firing as some components get started
later in the cluster run, might only have alerts in pending state. This
test hopes to catch those alerts that would potentially be in firing
state. But also out of the box we should not have any, ideally, alerts
in pending or firing states.

Lets see if this passes, if not will open bugzilla for each alert in pending state and add it to temp allowlist.

cc @wking

### CODE REVIEW

**Comment by lilic** [comment]
cc @openshift/openshift-team-monitoring

**Comment by wking** [comment]
Can we add this to the upgrade suite too?

**Comment by lilic** [comment]
> Can we add this to the upgrade suite too?

I would start with this first, and if it works we can do in a separate PR. sgty?

**Comment by wking** [comment]
Yeah, we could punt.  But we want the logic in a helper function, so we can easily use it in both places, right?

**Comment by lilic** [comment]
> But we want the logic in a helper function, so we can easily use it in both places, right?

It's a promql query, not sure it needs helper function for it?

**Comment by wking** [comment]
It's presumably going to end up with an exclusion list, right?

**Comment by wking** [comment]
Never mind, looks like the precedent set by #24786 is to have separate PromQL for upgrade and non-upgrade tests.  So I'm +1 to this as it stands (assuming CI doesn't turn up any existing pending alerts, as you said earlier).

File: test/extended/prometheus/prometheus.go [hunk 1]
```diff
@@ -88,6 +88,20 @@ var _ = g.Describe("[sig-instrumentation][Late] Alerts", func() {
...
+	g.It("should not have any alerts in pending state the entire cluster run", func() {
+		oc.SetupProject()
+		ns := oc.Namespace()
+		execPod := exutil.CreateCentosExecPodOrFail(oc.AdminKubeClient(), ns, "execpod", nil)
+		defer func() {
+			oc.AdminKubeClient().CoreV1().Pods(ns).Delete(context.Background(), execPod.Name, *metav1.NewDeleteOptions(1))
+		}()
+
+		tests := map[string]bool{
+			`count_over_time(ALERTS{alertstate="pending"}[2h])`: false,
```
Comment (paulfantom) [inline test/extended/prometheus/prometheus.go:100]: You need to exclude at least the same alerts as in https://github.com/openshift/origin/blob/master/test/extended/prometheus/prometheus.go#L69

Comment (paulfantom) [inline test/extended/prometheus/prometheus.go:100]: You need to exclude at least the same alerts as in https://github.com/openshift/origin/blob/master/test/extended/prometheus/prometheus.go#L69

File: test/extended/prometheus/prometheus.go [hunk 2]
```diff
@@ -88,6 +88,20 @@ var _ = g.Describe("[sig-instrumentation][Late] Alerts", func() {
 		e2e.Logf("Watchdog alert is firing")
 	})
 
+	g.It("should not have any alerts in pending state the entire cluster run", func() {
```
Comment (paulfantom) [inline test/extended/prometheus/prometheus.go:91]: Could you move it closer https://github.com/openshift/origin/blob/master/test/extended/prometheus/prometheus.go#L54 as those tests are very similar and need almost the same exclusion list?

File: test/extended/prometheus/prometheus.go [hunk 3]
```diff
@@ -88,6 +88,20 @@ var _ = g.Describe("[sig-instrumentation][Late] Alerts", func() {
...
+		oc.SetupProject()
+		ns := oc.Namespace()
+		execPod := exutil.CreateCentosExecPodOrFail(oc.AdminKubeClient(), ns, "execpod", nil)
+		defer func() {
+			oc.AdminKubeClient().CoreV1().Pods(ns).Delete(context.Background(), execPod.Name, *metav1.NewDeleteOptions(1))
+		}()
+
+		tests := map[string]bool{
+			`count_over_time(ALERTS{alertname!~"Watchdog|AlertmanagerReceiversNotConfigured|KubeAPILatencyHigh", alertstate="pending"}[2h])`: false,
```
Comment (wking) [inline test/extended/prometheus/prometheus.go:100]: Do we even need the `alertstate` filter here?  I'd rather have:

```go
`count_over_time(ALERTS{alertname!~"Watchdog|AlertmanagerReceiversNotConfigured|KubeAPILatencyHigh"}[2h])`: false
```

for "there are no surprising alerts in any state over the entire cluster run", to ensure we don't have something like a pending alert that slips through the firing test and then starts firing and slips through a pending-specific test.

**Comment by lilic** [comment]
Closing as it's not as easy as first thought as some alerts can be in pending state due to the nature of our e2e tests but don't eventually fire, so the list of allowing those would be too long. The general idea was to catch any alerts that are in pending state but eventually start firing, which sometimes our current tests miss, but it's not as easy. If someone manages to come up with an idea to make sure we catch all firing alerts, happy to review! 

/close

**Pull request closed without merging**